src/sentry/auth/providers/saml2.py

```diff
@@ -74,6 +74,36 @@ def dispatch(self, request, helper):
         return self.redirect(auth.login())
 
 
+# With SAML, the SSO request can be initiated by both the Service Provider
+# (sentry) (the typical case) and the Identity Provider. In the second case,
+# the auth assertion is directly posted to the ACS URL. Because the user will
+# not have initiated their SSO flow we must provide a endpoint similar to
+# auth_provider_login, but with support for initing the auth flow.
+class SAML2AcceptACSView(BaseView):
+    @method_decorator(csrf_exempt)
+    def dispatch(self, request, organization_slug):
+        # NB: The 'auth' key is used in the helper
+        in_auth_flow = request.session.get('auth', False)
+
+        # SP initiated authentication
+        if in_auth_flow:
+            from sentry.web.frontend.auth_provider_login import AuthProviderLoginView
+            sso_login = AuthProviderLoginView()
+            return sso_login.handle(request)
+
+        # IdP initiated authentication. Start from org login flow
+        from sentry.web.frontend.auth_organization_login import AuthOrganizationLoginView
+
+        # AuthOranizationLogin will init the login flow *only if* the ``init``
+        # parameter is set.
+        request.POST._mutable = True
+        request.POST['init'] = True
+        request.POST._mutable = False
```

Comment:
I'm not super happy about this at all, so definitely open to ideas

Reply:
🤢

idk what else you'd do unless you wanted to subclass some base "AuthClass" that just dry'd up the code

Reply:
There's not too much shared code between these views unfortunately, so it makes it difficult. I could break the `AuthOrganizationLoginView.handler` method into a separate method that takes an optional parameter to init the helper.

Reply:
Can't you copy into a new object then mutate that object, then reassign back to request.POST?

fwiw, I don't particularly care. I'd just maybe add a comment explaining this at least so it's an explicit decision.